src/Components/WebAssembly/WebAssembly/src/Services/LazyAssemblyLoader.cs

```diff
@@ -53,39 +55,47 @@ public async Task<IEnumerable<Assembly>> LoadAssembliesAsync(IEnumerable<string>
             return await LoadAssembliesInServerAsync(assembliesToLoad);
         }
 
-        private Task<IEnumerable<Assembly>> LoadAssembliesInServerAsync(IEnumerable<string> assembliesToLoad)
+        private async Task<IEnumerable<Assembly>> LoadAssembliesInServerAsync(IEnumerable<string> assembliesToLoad)
         {
-            var loadedAssemblies = _loadedAssemblyCache.Where(assembly =>
-                assembliesToLoad.Contains(assembly.GetName().Name + ".dll"));
+            var loadedAssemblies = new List<Assembly>();
 
-            if (loadedAssemblies.Count() != assembliesToLoad.Count())
+            try
             {
-                var unloadedAssemblies = assembliesToLoad.Except(loadedAssemblies.Select(a => a.GetName().Name + ".dll"));
-                throw new InvalidOperationException($"Unable to find the following assemblies: {string.Join(",", unloadedAssemblies)}. Make sure that the appplication is referencing the assemblies and that they are present in the output folder.");
+                foreach (var assemblyName in assembliesToLoad)
+                {
+                    loadedAssemblies.Add(Assembly.Load(Path.GetFileNameWithoutExtension(assemblyName)));
+                }
+            }
+            catch (FileNotFoundException ex)
+            {
+                throw new InvalidOperationException($"Unable to find the following assembly: {ex.FileName}. Make sure that the appplication is referencing the assemblies and that they are present in the output folder.");
             }
 
-            return Task.FromResult(loadedAssemblies);
+            return await Task.FromResult(loadedAssemblies);
         }
 
         private async Task<IEnumerable<Assembly>> LoadAssembliesInClientAsync(IEnumerable<string> assembliesToLoad)
         {
-            var jsRuntime = _provider.GetRequiredService<IJSRuntime>();
-            // Only load assemblies that haven't already been lazily-loaded
-            var newAssembliesToLoad = assembliesToLoad.Except(_loadedAssemblyCache.Select(a => a.GetName().Name + ".dll"));
+            // Check to see if the assembly has already been loaded and avoids reloading it if so.
+            // Note: in the future, as an extra precuation, we can call `Assembly.Load` and check
+            // to see if it throws FileNotFound to ensure that an assembly hasn't been loaded
+            // between when the cache of loaded assemblies was instantiated in the constructor
+            // and the invocation of this method.
+            var newAssembliesToLoad = assembliesToLoad.Except(_loadedAssemblyCache.Keys);
```

Comment:
Nit: This ends up [creating a `Set<string>` for the keys](https://github.com/dotnet/runtime/blob/847f2fa80fa47d3a579c6b08eeee8c774e20d3b7/src/libraries/System.Linq/src/System/Linq/Except.cs#L42-L43), but that seems unnecessary considering we already have a dictionary for lookup. It'd probably be better to use .Where(kvp => !oldDic.ContainsKey(kvp.Key))


```suggestion
            var newAssembliesToLoad = assembliesToLoad.Where(a => !_loadedAssemblyCache.ContainsKey(a));
```

Reply:
Good recommendation! Updated to use `Where`.